// react-map-gl-draw/dist-es6/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EDIT_TYPE = exports.ELEMENT_TYPE = exports.GUIDE_TYPE = exports.RENDER_STATE = exports.RENDER_TYPE = exports.SHAPE = exports.GEOJSON_TYPE = void 0;
var GEOJSON_TYPE;
exports.GEOJSON_TYPE = GEOJSON_TYPE;

(function (GEOJSON_TYPE) {
  GEOJSON_TYPE["POINT"] = "Point";
  GEOJSON_TYPE["LINE_STRING"] = "LineString";
  GEOJSON_TYPE["POLYGON"] = "Polygon";
})(GEOJSON_TYPE || (exports.GEOJSON_TYPE = GEOJSON_TYPE = {}));

var SHAPE;
exports.SHAPE = SHAPE;

(function (SHAPE) {
  SHAPE["POINT"] = "Point";
  SHAPE["LINE_STRING"] = "LineString";
  SHAPE["POLYGON"] = "Polygon";
  SHAPE["RECTANGLE"] = "Rectangle";
  SHAPE["CIRCLE"] = "Circle";
})(SHAPE || (exports.SHAPE = SHAPE = {}));

var RENDER_TYPE;
exports.RENDER_TYPE = RENDER_TYPE;

(function (RENDER_TYPE) {
  RENDER_TYPE["POINT"] = "Point";
  RENDER_TYPE["LINE_STRING"] = "LineString";
  RENDER_TYPE["POLYGON"] = "Polygon";
  RENDER_TYPE["RECTANGLE"] = "Rectangle";
})(RENDER_TYPE || (exports.RENDER_TYPE = RENDER_TYPE = {}));

var RENDER_STATE;
exports.RENDER_STATE = RENDER_STATE;

(function (RENDER_STATE) {
  RENDER_STATE["INACTIVE"] = "INACTIVE";
  RENDER_STATE["UNCOMMITTED"] = "UNCOMMITTED";
  RENDER_STATE["SELECTED"] = "SELECTED";
  RENDER_STATE["HOVERED"] = "HOVERED";
  RENDER_STATE["CLOSING"] = "CLOSING";
})(RENDER_STATE || (exports.RENDER_STATE = RENDER_STATE = {}));

var GUIDE_TYPE;
exports.GUIDE_TYPE = GUIDE_TYPE;

(function (GUIDE_TYPE) {
  GUIDE_TYPE["TENTATIVE"] = "tentative";
  GUIDE_TYPE["EDIT_HANDLE"] = "editHandle";
  GUIDE_TYPE["CURSOR_EDIT_HANDLE"] = "cursorEditHandle";
})(GUIDE_TYPE || (exports.GUIDE_TYPE = GUIDE_TYPE = {}));

var ELEMENT_TYPE;
exports.ELEMENT_TYPE = ELEMENT_TYPE;

(function (ELEMENT_TYPE) {
  ELEMENT_TYPE["FEATURE"] = "feature";
  ELEMENT_TYPE["FILL"] = "fill";
  ELEMENT_TYPE["SEGMENT"] = "segment";
  ELEMENT_TYPE["EDIT_HANDLE"] = "editHandle";
  ELEMENT_TYPE["TOOLTIP"] = "tooltip";
})(ELEMENT_TYPE || (exports.ELEMENT_TYPE = ELEMENT_TYPE = {}));

var EDIT_TYPE;
exports.EDIT_TYPE = EDIT_TYPE;

(function (EDIT_TYPE) {
  EDIT_TYPE["ADD_FEATURE"] = "addFeature";
  EDIT_TYPE["ADD_POSITION"] = "addPosition";
  EDIT_TYPE["REMOVE_POSITION"] = "removePosition";
  EDIT_TYPE["ADD_TENTATIVE_POSITION"] = "addTentativePosition";
  EDIT_TYPE["MOVE_POSITION"] = "movePosition";
  EDIT_TYPE["FINISH_MOVE_POSITION"] = "finishMovePosition";
})(EDIT_TYPE || (exports.EDIT_TYPE = EDIT_TYPE = {}));
