// react-map-gl-draw/dist/mode-handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _reactMapGl = require("react-map-gl");

var React = _interopRequireWildcard(require("react"));

var _editModes = require("@nebula.gl/edit-modes");

var _editingMode = _interopRequireDefault(require("./edit-modes/editing-mode"));

var _utils = require("./edit-modes/utils");

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultProps = {
  selectable: true,
  mode: null,
  features: null,
  onSelect: null,
  onUpdate: null,
  onUpdateCursor: function onUpdateCursor() {}
};
var defaultState = {
  featureCollection: new _editModes.ImmutableFeatureCollection({
    type: 'FeatureCollection',
    features: []
  }),
  selectedFeatureIndex: null,
  selectedEditHandleIndexes: [],
  // index, isGuide, mapCoords, screenCoords
  hovered: null,
  isDragging: false,
  didDrag: false,
  lastPointerMoveEvent: null,
  pointerDownPicks: null,
  pointerDownScreenCoords: null,
  pointerDownMapCoords: null
};

var ModeHandler = /*#__PURE__*/function (_React$PureComponent) {
  _inherits(ModeHandler, _React$PureComponent);

  var _super = _createSuper(ModeHandler);

  function ModeHandler(props) {
    var _this;

    _classCallCheck(this, ModeHandler);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "_events", void 0);

    _defineProperty(_assertThisInitialized(_this), "_eventsRegistered", void 0);

    _defineProperty(_assertThisInitialized(_this), "_modeHandler", void 0);

    _defineProperty(_assertThisInitialized(_this), "_context", void 0);

    _defineProperty(_assertThisInitialized(_this), "_containerRef", void 0);

    _defineProperty(_assertThisInitialized(_this), "getFeatures", function () {
      var featureCollection = _this._getFeatureCollection();

      featureCollection = featureCollection && featureCollection.getObject();
      return featureCollection && featureCollection.features;
    });

    _defineProperty(_assertThisInitialized(_this), "addFeatures", function (features) {
      var featureCollection = _this._getFeatureCollection();

      if (featureCollection) {
        if (!Array.isArray(features)) {
          features = [features];
        }

        featureCollection = featureCollection.addFeatures(features);

        _this.setState({
          featureCollection: featureCollection
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "deleteFeatures", function (featureIndexes) {
      var featureCollection = _this._getFeatureCollection();

      var selectedFeatureIndex = _this._getSelectedFeatureIndex();

      if (featureCollection) {
        if (!Array.isArray(featureIndexes)) {
          featureIndexes = [featureIndexes];
        }

        featureCollection = featureCollection.deleteFeatures(featureIndexes);
        var newState = {
          featureCollection: featureCollection
        };

        if (featureIndexes.findIndex(function (index) {
          return selectedFeatureIndex === index;
        }) >= 0) {
          newState.selectedFeatureIndex = null;
          newState.selectedEditHandleIndexes = [];
        }

        _this.setState(newState);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "deleteHandles", function (featureIndex, handleIndexes) {
      var featureCollection = _this._getFeatureCollection();

      if (!featureIndex) {
        featureIndex = _this._getSelectedFeatureIndex();
      }

      if (!handleIndexes) {
        if (!_this.state.selectedEditHandleIndexes.length) {
          return featureCollection;
        }

        handleIndexes = _this.state.selectedEditHandleIndexes;
      }

      var features = featureCollection.getObject().features; // It seems currently only POLYGON and LINE_STRING are supported
      // see handleClick event in editing-mode.ts

      var allowedTypes = [_constants.GEOJSON_TYPE.LINE_STRING, _constants.GEOJSON_TYPE.POLYGON];

      if (featureIndex !== null && features[featureIndex] && allowedTypes.includes(features[featureIndex].geometry.type)) {
        // Remove first indexes in DESC order
        handleIndexes.sort(function (n1, n2) {
          return n2 - n1;
        });
        var positionIndexes;

        if (features[featureIndex].geometry.type === _constants.GEOJSON_TYPE.LINE_STRING) {
          positionIndexes = handleIndexes.map(function (pos) {
            return [pos];
          });
        } else {
          // Currently only spport to handle simple polygons, thus pos 0
          positionIndexes = handleIndexes.map(function (pos) {
            return [0, pos];
          });
        }

        positionIndexes.forEach(function (pos) {
          featureCollection = featureCollection.removePosition(featureIndex, pos);
        });

        var selectedEditHandleIndexes = _this.state.selectedEditHandleIndexes.filter(function (handleIndex) {
          return !handleIndexes.includes(handleIndex);
        });

        _this.setState({
          featureCollection: featureCollection,
          selectedEditHandleIndexes: selectedEditHandleIndexes
        });
      }

      return featureCollection;
    });

    _defineProperty(_assertThisInitialized(_this), "_getMemorizedFeatureCollection", (0, _editModes._memoize)(function (_ref) {
      var propsFeatures = _ref.propsFeatures,
          stateFeatures = _ref.stateFeatures;
      var features = propsFeatures || stateFeatures; // Any changes in ImmutableFeatureCollection will create a new object

      if (features instanceof _editModes.ImmutableFeatureCollection) {
        return features;
      }

      if (features && features.type === 'FeatureCollection') {
        return new _editModes.ImmutableFeatureCollection({
          type: 'FeatureCollection',
          features: features.features
        });
      }

      return new _editModes.ImmutableFeatureCollection({
        type: 'FeatureCollection',
        features: features || []
      });
    }));

    _defineProperty(_assertThisInitialized(_this), "_getFeatureCollection", function () {
      return _this._getMemorizedFeatureCollection({
        propsFeatures: _this.props.features,
        stateFeatures: _this.state.featureCollection
      });
    });

    _defineProperty(_assertThisInitialized(_this), "_setupModeHandler", function () {
      var mode = _this.props.mode;
      _this._modeHandler = mode;

      if (!mode) {
        _this._degregisterEvents();

        return;
      }

      _this._registerEvents();
    });

    _defineProperty(_assertThisInitialized(_this), "_clearEditingState", function () {
      _this.setState({
        selectedFeatureIndex: null,
        selectedEditHandleIndexes: [],
        hovered: null,
        pointerDownPicks: null,
        pointerDownScreenCoords: null,
        pointerDownMapCoords: null,
        isDragging: false,
        didDrag: false
      });
    });

    _defineProperty(_assertThisInitialized(_this), "_getSelectedFeatureIndex", function () {
      if ('selectedFeatureIndex' in _this.props) {
        return _this.props.selectedFeatureIndex;
      }

      return _this.state.selectedFeatureIndex;
    });

    _defineProperty(_assertThisInitialized(_this), "_onSelect", function (selected) {
      var selectedFeatureIndex = selected.selectedFeatureIndex;
      var selectedEditHandleIndexes = _this.state.selectedEditHandleIndexes;
      var newState = {
        selectedFeatureIndex: selectedFeatureIndex,
        selectedEditHandleIndexes: selectedEditHandleIndexes
      };

      if (_this.state.selectedFeatureIndex !== selectedFeatureIndex) {
        newState.selectedEditHandleIndexes = [];
      }

      _this.setState(newState);

      if (_this.props.onSelect) {
        _this.props.onSelect(selected);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_onEdit", function (editAction) {
      var editType = editAction.editType,
          updatedData = editAction.updatedData,
          editContext = editAction.editContext;
      var newState = {
        featureCollection: new _editModes.ImmutableFeatureCollection(updatedData)
      };

      if (editType === _constants.EDIT_TYPE.ADD_POSITION) {
        // @ts-ignore
        newState.selectedEditHandleIndexes = [];
      }

      _this.setState(newState);

      switch (editType) {
        case _constants.EDIT_TYPE.ADD_FEATURE:
          _this._onSelect({
            selectedFeature: null,
            selectedFeatureIndex: null,
            selectedEditHandleIndex: null,
            selectedEditHandleIndexes: [],
            screenCoords: editContext && editContext.screenCoords,
            mapCoords: editContext && editContext.mapCoords
          });

          break;

        default:
      }

      if (_this.props.onUpdate) {
        _this.props.onUpdate({
          data: updatedData && updatedData.features,
          editType: editType,
          editContext: editContext
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_degregisterEvents", function () {
      var eventManager = _this._context && _this._context.eventManager;

      if (!_this._events || !eventManager) {
        return;
      }

      if (_this._eventsRegistered) {
        eventManager.off(_this._events);
        _this._eventsRegistered = false;
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_registerEvents", function () {
      var ref = _this._containerRef;
      var eventManager = _this._context && _this._context.eventManager;

      if (!_this._events || !ref || !eventManager) {
        return;
      }

      if (_this._eventsRegistered) {
        return;
      }

      eventManager.on(_this._events, ref);
      _this._eventsRegistered = true;
    });

    _defineProperty(_assertThisInitialized(_this), "_onEvent", function (handler, evt, stopPropagation) {
      var event = _this._getEvent(evt);

      handler(event);

      if (stopPropagation) {
        evt.stopImmediatePropagation();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_onClick", function (event) {
      var modeProps = _this.getModeProps(); // TODO refactor EditingMode
      // @ts-ignore


      if (_this._modeHandler instanceof _editingMode["default"] || _this.props.selectable) {
        var mapCoords = event.mapCoords,
            screenCoords = event.screenCoords;
        var pickedObject = event.picks && event.picks[0];

        var selectedEditHandleIndexes = _toConsumableArray(_this.state.selectedEditHandleIndexes); // @ts-ignore


        if (pickedObject && (0, _utils.isNumeric)(pickedObject.featureIndex)) {
          var handleIndex = // @ts-ignore
          pickedObject.type === _constants.ELEMENT_TYPE.EDIT_HANDLE ? pickedObject.index : null;
          var index = selectedEditHandleIndexes.indexOf(handleIndex);

          if (handleIndex !== null) {
            if (index !== -1) {
              selectedEditHandleIndexes.splice(index, 1);
            } else {
              selectedEditHandleIndexes.push(handleIndex);
            }

            _this.setState({
              selectedEditHandleIndexes: selectedEditHandleIndexes
            });
          } // @ts-ignore


          var selectedFeatureIndex = pickedObject.featureIndex;

          _this._onSelect({
            selectedFeature: pickedObject.object,
            selectedFeatureIndex: selectedFeatureIndex,
            selectedEditHandleIndex: handleIndex,
            selectedEditHandleIndexes: selectedEditHandleIndexes,
            // @ts-ignore
            mapCoords: mapCoords,
            screenCoords: screenCoords
          });
        } else {
          _this._onSelect({
            selectedFeature: null,
            selectedFeatureIndex: null,
            selectedEditHandleIndex: null,
            selectedEditHandleIndexes: selectedEditHandleIndexes,
            // @ts-ignore
            mapCoords: mapCoords,
            screenCoords: screenCoords
          });
        }
      }

      _this._modeHandler.handleClick(event, modeProps);
    });

    _defineProperty(_assertThisInitialized(_this), "_onDblclick", function (event) {
      if ((0, _utils.isNumeric)(_this._getSelectedFeatureIndex())) {
        event.sourceEvent.stopImmediatePropagation();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_onPointerMove", function (event) {
      // hovering
      var hovered = _this._getHoverState(event);

      var _this$state = _this.state,
          isDragging = _this$state.isDragging,
          didDrag = _this$state.didDrag,
          pointerDownPicks = _this$state.pointerDownPicks,
          pointerDownScreenCoords = _this$state.pointerDownScreenCoords,
          pointerDownMapCoords = _this$state.pointerDownMapCoords;

      if (isDragging && !didDrag && pointerDownScreenCoords) {
        var dx = event.screenCoords[0] - pointerDownScreenCoords[0];
        var dy = event.screenCoords[1] - pointerDownScreenCoords[1];

        if (dx * dx + dy * dy > 5) {
          _this.setState({
            didDrag: true
          });
        }
      }

      var pointerMoveEvent = _objectSpread({}, event, {
        isDragging: isDragging,
        pointerDownPicks: pointerDownPicks,
        pointerDownScreenCoords: pointerDownScreenCoords,
        pointerDownMapCoords: pointerDownMapCoords,
        cancelPan: event.sourceEvent.stopImmediatePropagation
      });

      var modeProps = _this.getModeProps();

      _this._modeHandler.handlePointerMove(pointerMoveEvent, modeProps);

      if (_this.state.didDrag) {
        _this._modeHandler.handleDragging(pointerMoveEvent, modeProps);
      }

      _this.setState({
        hovered: hovered,
        // @ts-ignore
        lastPointerMoveEvent: pointerMoveEvent
      });
    });

    _defineProperty(_assertThisInitialized(_this), "_onPointerDown", function (event) {
      var dragToDraw = _this.props.modeConfig && _this.props.modeConfig.dragToDraw;
      var isDragging = Boolean(event.picks && event.picks[0]) || dragToDraw;

      var startDraggingEvent = _objectSpread({}, event, {
        isDragging: isDragging,
        pointerDownScreenCoords: event.screenCoords,
        pointerDownMapCoords: event.mapCoords,
        cancelPan: event.sourceEvent.stopImmediatePropagation
      });

      var newState = {
        isDragging: isDragging,
        pointerDownPicks: event.picks,
        pointerDownScreenCoords: event.screenCoords,
        pointerDownMapCoords: event.mapCoords
      }; // @ts-ignore

      _this.setState(newState);

      var modeProps = _this.getModeProps();

      _this._modeHandler.handleStartDragging(startDraggingEvent, modeProps);
    });

    _defineProperty(_assertThisInitialized(_this), "_onPointerUp", function (event) {
      var _this$state2 = _this.state,
          didDrag = _this$state2.didDrag,
          pointerDownPicks = _this$state2.pointerDownPicks,
          pointerDownScreenCoords = _this$state2.pointerDownScreenCoords,
          pointerDownMapCoords = _this$state2.pointerDownMapCoords;

      var stopDraggingEvent = _objectSpread({}, event, {
        isDragging: false,
        pointerDownPicks: didDrag ? pointerDownPicks : null,
        pointerDownScreenCoords: didDrag ? pointerDownScreenCoords : null,
        pointerDownMapCoords: didDrag ? pointerDownMapCoords : null,
        cancelPan: event.sourceEvent.cancelPan
      });

      var newState = {
        isDragging: false,
        didDrag: false,
        pointerDownPicks: null,
        pointerDownScreenCoords: null,
        pointerDownMapCoords: null
      };

      _this.setState(newState);

      var modeProps = _this.getModeProps();

      if (didDrag) {
        _this._modeHandler.handleStopDragging(stopDraggingEvent, modeProps);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_onPan", function (event) {
      var isDragging = _this.state.isDragging;

      if (isDragging) {
        event.sourceEvent.stopImmediatePropagation();
      }

      if (_this._modeHandler.handlePan) {
        _this._modeHandler.handlePan(event, _this.getModeProps());
      }
    });

    _defineProperty(_assertThisInitialized(_this), "project", function (pt) {
      var viewport = _this._context && _this._context.viewport;
      return viewport && viewport.project(pt);
    });

    _defineProperty(_assertThisInitialized(_this), "unproject", function (pt) {
      var viewport = _this._context && _this._context.viewport;
      return viewport && viewport.unproject(pt);
    });

    _defineProperty(_assertThisInitialized(_this), "_getHoverState", function (event) {
      var object = event.picks && event.picks[0];

      if (!object) {
        return null;
      }

      return _objectSpread({
        screenCoords: event.screenCoords,
        mapCoords: event.mapCoords
      }, object);
    });

    _this.state = defaultState;
    _this._eventsRegistered = false;
    _this._events = {
      anyclick: function anyclick(evt) {
        return _this._onEvent(_this._onClick, evt, true);
      },
      dblclick: function dblclick(evt) {
        return _this._onEvent(_this._onDblclick, evt, false);
      },
      click: function click(evt) {
        return evt.stopImmediatePropagation();
      },
      pointermove: function pointermove(evt) {
        return _this._onEvent(_this._onPointerMove, evt, false);
      },
      pointerdown: function pointerdown(evt) {
        return _this._onEvent(_this._onPointerDown, evt, true);
      },
      pointerup: function pointerup(evt) {
        return _this._onEvent(_this._onPointerUp, evt, true);
      },
      panmove: function panmove(evt) {
        return _this._onEvent(_this._onPan, evt, false);
      },
      panstart: function panstart(evt) {
        return _this._onEvent(_this._onPan, evt, false);
      },
      panend: function panend(evt) {
        return _this._onEvent(_this._onPan, evt, false);
      }
    };
    return _this;
  }

  _createClass(ModeHandler, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this._setupModeHandler();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (prevProps.mode !== this.props.mode) {
        this._clearEditingState();

        this._setupModeHandler();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._degregisterEvents();
    }
  }, {
    key: "getModeProps",
    value: function getModeProps() {
      var featureCollection = this._getFeatureCollection();

      var lastPointerMoveEvent = this.state.lastPointerMoveEvent;

      var selectedFeatureIndex = this._getSelectedFeatureIndex();

      var selectedEditHandleIndexes = this.state.selectedEditHandleIndexes;
      var viewport = this._context && this._context.viewport;
      return {
        data: featureCollection && featureCollection.featureCollection,
        selectedIndexes: (0, _utils.isNumeric)(selectedFeatureIndex) ? [selectedFeatureIndex] : [],
        selectedEditHandleIndexes: selectedEditHandleIndexes,
        lastPointerMoveEvent: lastPointerMoveEvent,
        viewport: viewport,
        featuresDraggable: this.props.featuresDraggable,
        onEdit: this._onEdit,
        onUpdateCursor: this.props.onUpdateCursor,
        modeConfig: this.props.modeConfig
      };
    }
    /* MEMORIZERS */

  }, {
    key: "_getEvent",
    value: function _getEvent(evt) {
      var features = this.getFeatures();

      var guides = this._modeHandler.getGuides(this.getModeProps());

      var picked = (0, _utils.parseEventElement)(evt, features, guides && guides.features);
      var screenCoords = (0, _utils.getScreenCoords)(evt); // @ts-ignore

      var mapCoords = this.unproject(screenCoords);
      return {
        picks: picked ? [picked] : null,
        screenCoords: screenCoords,
        mapCoords: mapCoords,
        sourceEvent: evt
      };
    }
  }, {
    key: "_render",
    value: function _render() {
      return /*#__PURE__*/React.createElement("div", null);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return /*#__PURE__*/React.createElement(_reactMapGl._MapContext.Consumer, null, function (context) {
        _this2._context = context;
        var viewport = context && context.viewport;

        if (!viewport || viewport.height <= 0 || viewport.width <= 0) {
          return null;
        }

        return _this2._render();
      });
    }
  }]);

  return ModeHandler;
}(React.PureComponent);

exports["default"] = ModeHandler;

_defineProperty(ModeHandler, "displayName", 'ModeHandler');

_defineProperty(ModeHandler, "defaultProps", defaultProps);
